File: JavaScript/1366-Rank Teams by Votes.js
/**
 * @param {string[]} votes
 * @return {string}
 */
var rankTeams = function(votes) {
    let hmap = new Map();
    const N = votes[0].length;
    
    for (let word of votes[0]){
        hmap.set(word, new Array(N).fill(0) );
    }
    
    for (let vote of votes){
        for (let i=0; i<N; i++){
            hmap.get(vote[i])[i] += 1;
        }
    }
    
    let res = [...votes[0]];

    res.sort((a,b)=>{
        for (i=0; i<N; i++){
            if (hmap.get(a)[i] !== hmap.get(b)[i]){
                return hmap.get(b)[i] - hmap.get(a)[i];
            }
        }
        return a.charCodeAt(0) - b.charCodeAt(0);
    })
    
    return res.join("")
};

File: 1366-Rank Teams by Votes.py
class Solution(object):
    def rankTeams(self, votes):
        """
        :type votes: List[str]
        :rtype: str
        """
# Time O(NM) for iterating
# Time O(MMlogM) for sorting
# Space O(M^2)
# where N = votes.length and M = votes[0].length <= 26
        
        N = len(votes[0])
        hmap = {c: [0]*N + [-ord(c)] for c in votes[0]}
        
        for vote in votes:
            for i, c in enumerate(vote):
                hmap[c][i] += 1

        res = list(votes[0])
        res.sort(key = lambda x: hmap[x], reverse = True)
        
        return "".join(res)